### PR TITLE
[Feature] Automated globus transfer

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
Finally got this working - tests are passing on any local device logged into globus.

Depends on the process management helpers in #492, which I've now made non-private since it's proving quite useful generally. (@h-mayorquin should be happy to hear it uses `subprocess` instead of `os`)

This is the final item required to fully automated the Tingley conversion, which I'll start doing now.

### CODE REVIEW

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 1]
```diff
@@ -61,14 +63,143 @@ def get_globus_dataset_content_sizes(
...
+        deploy_process(
             command=f"globus ls -Fjson {globus_endpoint_id}:{path}{recursive_flag}", catch_output=True, timeout=timeout
         )
     )
     files_and_sizes = {item["name"]: item["size"] for item in contents["DATA"] if item["type"] == "file"}
     return files_and_sizes
 
 
+def transfer_globus_content(
+    source_id: str,
```
Comment (bendichter) [inline src/nwb_conversion_tools/tools/data_transfers.py:75]: would you mind this change to make it a bit more explicit?

```suggestion
    source_globus_id: str,
```

Comment (bendichter) [inline src/nwb_conversion_tools/tools/data_transfers.py:75]: would you mind this change to make it a bit more explicit?

```suggestion
    source_globus_id: str,
```

Comment (bendichter) [reply src/nwb_conversion_tools/tools/data_transfers.py:75]: or source_endpoint_id

Comment (bendichter) [reply src/nwb_conversion_tools/tools/data_transfers.py:75]: or source_endpoint_id

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 2]
```diff
@@ -61,14 +63,143 @@ def get_globus_dataset_content_sizes(
...
+    for j, batched_source_files in enumerate(source_files):
+        paths_file = Path(destination_folder) / f"paths_{j}.txt"
+        # .as_posix() to ensure correct string form Globus expects
+        # ':' replacement for Windows drives
+        source_folder = Path(batched_source_files[0]).parent.as_posix().replace(":", "")
+        destination_folder_name = Path(destination_folder).as_posix().replace(":", "")
+        with open(file=paths_file, mode="w") as f:
+            for source_file in batched_source_files:
+                file_name = Path(source_file).name
+                f.write(f"{file_name} {file_name}\n")
```
Comment (bendichter) [inline src/nwb_conversion_tools/tools/data_transfers.py:146]: The easiest way to pass files in is to create an intermediate text file?

Comment (bendichter) [inline src/nwb_conversion_tools/tools/data_transfers.py:146]: The easiest way to pass files in is to create an intermediate text file?

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 3]
```diff
@@ -61,14 +63,143 @@ def get_globus_dataset_content_sizes(
...
+        Maximum amount of time to monitor the transfer progress.
+        You may wish to set this to be longer when transferring very large files.
+        Defaults to 10 minutes.
+
+    Returns
+    -------
+    success : bool
+        If 'display_progress'=False, and the transfer was succesfully initiated, then this function returns True.
+
+        If 'display_progress'=True (the default), then this function returns the total status of all transfers
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:120]:
```suggestion
        If 'display_progress'=True (the default), then this function returns the total size of all transfers
```
Is this what you mean?

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 4]
```diff
@@ -61,14 +63,143 @@ def get_globus_dataset_content_sizes(
...
+            ),
+            string=transfer_message,
+        )
+        paths_file.unlink()
+        assert task_id, f"Transfer submission failed! Globus output:\n{transfer_message}."
+
+        if source_folder not in folder_content_sizes:
+            contents = get_globus_dataset_content_sizes(globus_endpoint_id=source_id, path=source_folder)
+            folder_content_sizes.update({source_folder: contents})
+        task_total_sizes.update(
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py:169]: It seems a little bit complicated to see what structure you get at the end. Flagging it so I can think a little bit better about it. Maybe there is a better way of writing this.

File: tests/test_internals/test_tools.py [hunk 5]
```diff
@@ -181,3 +182,34 @@ def tearDown(self):
...
+        ]
+        success, task_ids = transfer_globus_content(
+            source_id=source_id,
+            source_files=test_source_files,
+            destination_id=destination_id,
+            destination_folder=self.tmpdir,
+            display_progress=False,
+        )
+        assert success
+        assert task_ids
```
Comment (h-mayorquin) [inline tests/test_internals/test_tools.py:215]: Is this just testing that the function does not return empty arguments? Can we test size file in the temporary folder?

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 2 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:146]: The *only* way to use the `--batch` option, regrettably. https://github.com/globus/globus-cli/blob/main/src/globus_cli/commands/transfer.py#L155-L165

Could technically go down to that lower level of the `globus_sdk` library and add the items programmatically instead, but that requires a heck of a lot of other steps to get unpacked and setup before doing that (auth/client/every other option). Would be much harder to maintain on our end.

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:120]: Maybe just 'progress' if 'status' is too equivocal with being boolean.

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:169]: double-nested dict tree - outer level is unique common parent folders, inner level is filename in that folder with value equal to size in bytes.

Not that the JSON content API request is that expensive, but still, you'd really only want it to have to grab the info once per unique folder.

File: tests/test_internals/test_tools.py (hunk 5 above)
Comment (CodyCBakerPhD) [reply tests/test_internals/test_tools.py:215]: `success` is `bool`, so asserts that it is True.

`task_ids` is a list, so asserts that it is not empty.

> Can we test size file in the temporary folder?

Yeah, sure.

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:120]: No, 'status' is correct, everything here is boolean. `success` is the global status, `all(all_status)`, and `all_status` is the list of booleans.

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 6]
```diff
@@ -61,14 +63,143 @@ def get_globus_dataset_content_sizes(
...
 
+def transfer_globus_content(
+    source_id: str,
+    source_files: Union[str, List[List[str]]],
+    destination_id: str,
+    destination_folder: FolderPathType,
+    display_progress: bool = True,
+    progress_update_rate: float = 60.0,
+    progress_update_timeout: float = 600.0,
+) -> (bool, List[str]):
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/data_transfers.py]: This is flagged in my editor, shouldn't it be `Tuple(bool, List[str])`?

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/data_transfers.py]: I see, so maybe: then this functions returns success=True if all transfer have succeeded?
I think what confuses me is the phrase "returns total status" by which you mean returns `all(all_status)`

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py]: Done

File: tests/test_internals/test_tools.py (hunk 5 above)
Comment (CodyCBakerPhD) [reply tests/test_internals/test_tools.py:215]: Thanks for pointing this out, actually revealed an issue with the relative globus paths. Working on it now.

**Review by h-mayorquin — COMMENTED** [review]
Shouldn't this be two functions? 
The first puts forward the globus process and returns one argument that is the structure of what is to be transfered: the dictionary "task_total_sizes" (it returns an empty dict if it fails).
(lines 125-179)
The second gets that dictionary and monitors the progress as it occurs. Returns True if everything went well, returns false if anything failed in the middle.
(lines 180-200)

Those functions would have a more clear responsibility.

**Comment by CodyCBakerPhD** [comment]
> Shouldn't this be two functions?
The first puts forward the globus process and returns one argument that is the structure of what is to be transfered: the dictionary "task_total_sizes" (it returns an empty dict if it fails).
(lines 125-179)
The second gets that dictionary and monitors the progress as it occurs. Returns True if everything went well, returns false if anything failed in the middle.
(lines 180-200)
>
> Those functions would have a more clear responsibility.

I can make them private sub-functions if you think that would be better. The real trouble is that subfunction 1 is asynchronous but we want to force it to lock the running of a script that would then run the full data conversion the moment the transfer completes (assuming `success=True`, else need manual intervention). So they do need to go hand-in-hand for the primary function call.

File: src/nwb_conversion_tools/tools/data_transfers.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py]: > I see, so maybe: then this functions returns success=True if all transfer have succeeded?

Correct.

> I think what confuses me is the phrase "returns total status" by which you mean returns all(all_status)

Yeah, 'total status' = combined success of all batches

**Comment by h-mayorquin** [comment]
> > Shouldn't this be two functions?
> > The first puts forward the globus process and returns one argument that is the structure of what is to be transfered: the dictionary "task_total_sizes" (it returns an empty dict if it fails).
> > (lines 125-179)
> > The second gets that dictionary and monitors the progress as it occurs. Returns True if everything went well, returns false if anything failed in the middle.
> > (lines 180-200)
> > Those functions would have a more clear responsibility.
> 
> I can make them private local sub-functions if you think that would be better. The real trouble is that subfunction 1 is asynchronous but we want to force it to lock continuation of a script that would then run the full data conversion the moment the transfer completes (assuming `success=True`, else need manual intervention). So they do need to go hand-in-hand for the primary function call.

I think that making them private local sub-functions would at least help from a readability point of view. For the case that you describe, could you not lock the start of the conversion to the success of the hypothetical sub_function 2?

**Comment by CodyCBakerPhD** [comment]
> I think that making them private local sub-functions would at least help from a readability point of view. For the case that you describe, could you not lock the start of the conversion to the success of the hypothetical sub_function 2? the second function would be a tracking a function (with the option of displaying progress) which will liberate the script once all the transfers are succesfully.

Gotta remember to do that though from an API point of view. Much safer to not leave room for that error on the user side (mostly my side, in this case). This also puts the behavior closer to the recently added figshare download tools.

**Comment by h-mayorquin** [comment]
> > I think that making them private local sub-functions would at least help from a readability point of view. For the case that you describe, could you not lock the start of the conversion to the success of the hypothetical sub_function 2? the second function would be a tracking a function (with the option of displaying progress) which will liberate the script once all the transfers are succesfully.
> 
> Gotta remember to do that though from an API point of view. Much safer to not leave room for that error on the user side (mostly my side, in this case). This also puts the behavior closer to the recently added figshare download tools.

Fair enough, I still favor the approach of using two internal private methods or at least adding comments at the beginning of each of the indicated sections for improving readability. 

Something that surfaced form discussion, it seems that you are tying your locking mechanism to display progress. If the function needs to be locked in some specific context maybe you want to separate display from locking? I think that the function will be easier to document with that distinction.

**Comment by CodyCBakerPhD** [comment]
> Fair enough, I still favor the approach of using two internal private methods or at least adding comments at the beginning of each of the indicated sections for improving readability.

Already done!

>  it seems that you are tying your locking mechanism to display progress. If the function needs to be locked in some specific context maybe you want to separate display from locking? I think that the function will be easier to document with that distinction.

Are you sure about that?

**Comment by CodyCBakerPhD** [comment]
> Are you sure about that?

Trick question, I started it out that way but realized exactly what you are saying, so I fixed it. 😉

**Comment by h-mayorquin** [comment]
> > Are you sure about that?
> 
> Trick question, I started it out that way but realized exactly what you are saying, so I fixed it. wink

: D
Yep, I looked at the old function and I had that in my mind when I was writing. 

Two minor details: 
1) The first private method is using the outer scope variable `destination_folder_path` instead of the local one `destination_folder` inside of the first sub_function. I think that the outer scope name is better btw. 
2) The second private method is usign the outer scope `task_total_sizes` instead of the intended `tasks_ids` which is not used.

**Comment by CodyCBakerPhD** [comment]
> Two minor details:
>
> The first private method is using the outer scope variable destination_folder_path instead of the local one destination_folder inside of the first sub_function. I think that the outer scope name is better btw.
The second private method is usign the outer scope task_total_sizes instead of the intended tasks_ids which is not used.

Good catches! Thanks

**Comment by bendichter** [comment]
Why are `_submit_transfer_request` and `_track_transfer` defined within the scope of `transfer_globus_content`?

**Comment by CodyCBakerPhD** [comment]
> Why are _submit_transfer_request and _track_transfer defined within the scope of transfer_globus_content?

See discussion with Heberto, who requested this for readability

**Comment by bendichter** [comment]
I feel like marking the functions as private with a leading underscore is good enough. I generally only define a function within another function if I want to share scope. But this is more of a stylistic thing, and if this way is easier for you to understand then I think it's fine

**Review by bendichter — APPROVED** [review]

File: src/nwb_conversion_tools/tools/data_transfers.py [hunk 7]
```diff
@@ -61,14 +63,175 @@ def get_globus_dataset_content_sizes(
...
+        source_endpoint_id: str,
+        source_files: Union[str, List[List[str]]],
+        destination_endpoint_id: str,
+        destination_folder_path: Path,
+    ) -> Dict[str, int]:
+        """Send transfer request to Globus."""
+        folder_content_sizes = dict()
+        task_total_sizes = dict()
+        for j, batched_source_files in enumerate(source_files):
+            paths_file = destination_folder_path / f"paths_{j}.txt"
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb_conversion_tools/tools/data_transfers.py:133]: I tried that at first however you are not guaranteed to have write permissions there. The function can only be guaranteed to have write permission at the destination folder path. I mean, I could make a temporary sub-directory there but the effect would be the same

**Comment by CodyCBakerPhD** [comment]
> I feel like marking the functions as private with a leading underscore is good enough. I generally only define a function within another function if I want to share scope. But this is more of a stylistic thing, and if this way is easier for you to understand then I think it's fine

It's entirely for style - I'm personally fine with not even using sub-functions at all, but @h-mayorquin found it stylistically easier to understand.

It is perhaps conceivable that I might eventually want the task tracking to be available as a global function, I'll have to see after getting some practical use out of this. But I could also see a class-based approach being the superior API for all of this, so rearranging stuff would be best done then.

**Comment by h-mayorquin** [comment]
> > I feel like marking the functions as private with a leading underscore is good enough. I generally only define a function within another function if I want to share scope. But this is more of a stylistic thing, and if this way is easier for you to understand then I think it's fine

I actually would have preferred to have the functions as independent private methods. I somehow misunderstood that this was not doable for locking reasons.

**Pull request merged**